Don't relocate Apple Silicon bottles for default prefix

samuelarogbonlo opened the pull request:
This change skips the relocation process for bottles on Apple Silicon Macs when using the default prefix (/opt/homebrew). Relocation was initially needed for Intel Macs to avoid wide-reaching replacements with false positives, but is unnecessary for Apple Silicon where the default prefix is unique.

Benefits:
- Simplifies and speeds up bottle pouring on Apple Silicon
- Potentially enables code-signing of homebrew-core packages
- Improves security by avoiding binary modification post-build

A --force-bottle-relocation flag is added for edge cases where relocation might still be needed.

Fixes #19247

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?


Note: This PR is modifying autogenerated files that differ between Apple Silicon and Linux environments. Please run `brew generate-man-completions` during merge to ensure proper file generation.

-----

samuelarogbonlo commented:
Based on the feedback, I've updated the implementation to:

1. Remove the --force-bottle-relocation flag as suggested
2. Add an environment variable (HOMEBREW_BOTTLE_SKIP_RELOCATION_ARM64) for gradual rollout
3. Default to the current behavior (no skipping) until the feature is explicitly enabled

For the next steps, I'd like to address the remaining suggestions:

1. Implement binary file relocation detection using RubyMacho to check if binaries actually need relocation before skipping
2. Store whether relocation was skipped in the bottle manifest/tab
3. Consider moving bottle relocation from bottle-creation time to bottle-pour time for arm64 as suggested by @carlocab

I'd appreciate some guidance on how to best implement these remaining items, particularly around using RubyMacho for analyzing binaries, and the preferred approach for tracking relocation status in the bottle metadata. Are there examples in the existing codebase I should look at for reference?

carlocab commented:
Hey @samuelarogbonlo, thanks for giving this a go!

I don't think I'll have answers to all your questions, but I'll try to give you more detailed guidance in a few days. Feel free to ping me this weekend as a reminder.

samuelarogbonlo commented:
> Hey @samuelarogbonlo, thanks for giving this a go!
> 
> I don't think I'll have answers to all your questions, but I'll try to give you more detailed guidance in a few days. Feel free to ping me this weekend as a reminder.


@carlocab This is a friendly reminder. Thanks

MikeMcQuaid commented:
> I'd appreciate some guidance on how to best implement these remaining items, particularly around using RubyMacho for analyzing binaries, and the preferred approach for tracking relocation status in the bottle metadata. Are there examples in the existing codebase I should look at for reference?

@samuelarogbonlo You can look at the relocation logic in `keg_relocate.rb` files and `bottle.rb`, install receipt/tab logic in `tab.rb` and `formula_installer.rb` for pour-time logic. Hope that helps!

samuelarogbonlo commented:
I've updated the PR based on the feedback:

1. Removed the --force-bottle-relocation flag
2. Added an environment variable (HOMEBREW_BOTTLE_SKIP_RELOCATION_ARM64) for gradual rollout
3. Implemented binary file relocation detection using RubyMacho
4. Added tracking of relocation status in the bottle tab metadata

The implementation now: 
- Only skips relocation when it's truly unnecessary (no binary files need relocation)
- Provides a controlled rollout mechanism via environment variable
- Records whether relocation was skipped for transparency

I looked at the relocation logic in keg_relocate.rb and bottle.rb, and the tab logic in tab.rb and formula_installer.rb as suggested. Let me know if there are other improvements you'd like to see!

carlocab commented on Library/Homebrew/extend/os/mac/utils/bottles.rb:
```diff
@@ -11,6 +11,11 @@ def tag(tag = nil)
           super
         end
+      
+        sig {returns(T::Boolean) }
+        def on_macos?
+          true
+        end
```
Instead of defining an `on_macos?` method (which is essentially the same thing as checking for `OS.mac?`), instead, implement each method (e.g. `binaries_need_relocation?` as needed inside each of `extend/os/{mac,linux}/utils/bottles.rb`.

carlocab commented on Library/Homebrew/formula_installer.rb:
```diff
@@ -1502,6 +1502,7 @@ def pour
     tab.unused_options = []
     tab.built_as_bottle = true
     tab.poured_from_bottle = true
+    if Hardware::CPU.arm? && Utils::Bottles.on_macos?
```
This `if` block seems a bit weirdly placed. Is what's under it indented correctly?

carlocab commented on Library/Homebrew/utils/bottles.rb:
```diff
@@ -128,6 +128,44 @@ def load_tab(formula)
+
+        keg.mach_o_files.any? do |file|
+          # Check if dylib ID contains paths that need relocation
+          (file.dylib? && file.dylib_id&.include?("/usr/local")) ||
+          # Check if linked libraries contain paths that need relocation
+          file.dynamically_linked_libraries.any? { |lib| lib.include?("/usr/local") }
```
Shouldn't we be checking for `HOMEBREW_MACOS_ARM_DEFAULT_PREFIX` here?

carlocab reviewed (commented):
Apologies for the delayed response here; life/work gets in the way.

I think Mike gave you a few places to look at above for inspiration. I also have a suggested alternative approach:

Instead of making `replace_locations_with_placeholders` and `replace_placeholders_with_locations` do nothing on Apple Silicon:
- make them do nothing if you are on Apple Silicon AND using the default prefix of `/opt/homebrew`
- adjust the [placeholders](https://github.com/Homebrew/brew/blob/b4fe541766c3788c2bd9542a4d35c69d108a1590/Library/Homebrew/keg_relocate.rb#L5-L10) so that on Apple Silicon, they refer to concrete paths (e.g. `/opt/homebrew` for `@@HOMEBREW_PREFIX@@`, etc)

Doing the second should hopefully get you most of the way towards making it so that relocation effectively happens at bottle-pour time on Apple Silicon when in a non-default prefix (which is what I suggested [here](https://github.com/Homebrew/brew/pull/19384#issuecomment-2685598102)).

Thanks again for persevering here! Let me know if you have any questions. (I'll try to be more responsive next time 😅)

MikeMcQuaid commented:
> * make them do nothing if you are on Apple Silicon AND using the default prefix of `/opt/homebrew`
> * adjust the [placeholders](https://github.com/Homebrew/brew/blob/b4fe541766c3788c2bd9542a4d35c69d108a1590/Library/Homebrew/keg_relocate.rb#L5-L10) so that on Apple Silicon, they refer to concrete paths (e.g. `/opt/homebrew` for `@@HOMEBREW_PREFIX@@`, etc)

Seems sensible 👍🏻.

The most important thing here is that the files are not rewritten/have checksums changed on Apple Silicon.

samuelarogbonlo commented:
Thank you for the feedback! I've updated the implementation to:

1. Only skip relocation on Apple Silicon with the default prefix
2. Move method implementations to platform-specific files
3. Check for HOMEBREW_MACOS_ARM_DEFAULT_PREFIX in binary detection
4. Fix indentation in formula_installer.rb

This approach allows for both performance benefits on default installations and proper relocation on custom prefix installations.
